pyccel/ast/numpyext.py

```diff
@@ -675,7 +675,7 @@ class NumpyFullLike:
     def __new__(cls, a, fill_value, dtype=None, order='K', subok=True, shape=None):
 
         # NOTE: we ignore 'subok' argument
-        dtype = a.dtype if (dtype is None) or isinstance(dtype, Nil) else dtype
+        dtype = a.dtype if not dtype else dtype
```

Comment:
This kind of statements can be written in a completely equivalent for,
```
dtype = dtype or a.dtype
```
because the Python `or` operator returns the first argument that is not False.

See [here](https://stackoverflow.com/questions/4978738/is-there-a-python-equivalent-of-the-c-sharp-null-coalescing-operator) for example.

Reply:
I didn't realise we could do this! Does this mean we need to open an issue about pyccel's `or` operator? Our translations always give the result of `or` as a boolean

Reply:
I'm not sure that we can mimic Python's behavior with Pyccel... but we can try :wink:

Reply:
I think it should be possible, we just need to translate:
```
a or b
```
to:
```
a if bool(a) else b
```
if `a` and `b` are not booleans. This can easily be done in the `PyccelOr.__new__`

Reply:
why did you change that , imagine if the user writes, somthing like `a = np.array([1,2,3],dtype=None)`

Reply:
@saidctb The behaviour is the same as this PR ensures that `bool(Nil())` returns `False`

Reply:
ah okay